pkg/codegen/hcl2/model/type_const.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2016-2021, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/pulumi/pulumi/pkg/v3/codegen/hcl2/syntax"
+)
+
+// ConstType represents a type that is a single constant value.
+type ConstType struct {
+	// Type is the underlying value type.
+	Type Type
+	// Value is the constant value.
+	Value interface{}
+}
+
+// NewConstType creates a new constant type with the given type and value.
+func NewConstType(typ Type, value interface{}) *ConstType {
+	return &ConstType{Type: typ, Value: value}
+}
+
+// SyntaxNode returns the syntax node for the type. This is always syntax.None.
+func (*ConstType) SyntaxNode() hclsyntax.Node {
+	return syntax.None
+}
+
+// Traverse attempts to traverse the type with the given traverser. The result is the traversal
+// result of the underlying type.
+func (t *ConstType) Traverse(traverser hcl.Traverser) (Traversable, hcl.Diagnostics) {
+	return t.Type.Traverse(traverser)
+}
+
+// Equals returns true if this type has the same identity as the given type.
+func (t *ConstType) Equals(other Type) bool {
+	return t.equals(other, nil)
+}
+
+func (t *ConstType) equals(other Type, seen map[Type]struct{}) bool {
+	if t == other {
+		return true
+	}
+
+	otherConst, ok := other.(*ConstType)
+	return ok && t.Value == otherConst.Value && t.Type.equals(otherConst.Type, seen)
+}
+
+// AssignableFrom returns true if this type is assignable from the indicated source type. A token(name) is assignable
```

Comment:
Could you file an issue for follow-up to expand convertibility and assignability for these types? E.g. a constant string should be convertible to a string (which requires changes in the core convertibility rules).

Also, `s/token(name)/const(value)/g` :)

Reply:
https://github.com/pulumi/pulumi/issues/6967